### PR TITLE
Fixes #15634 - RancherOS support

### PR DESCRIPTION
**Pull request by selmison**
Add support to RancherOS family.

[https://community.theforeman.org/t/support-rancheros/8161](https://community.theforeman.org/t/support-rancheros/8161)

### CODE REVIEW

**Comment by selmison** [comment]
Sample PXE template used in my environment:

```<%#
kind: PXELinux
name: RancherOS PXELinux
-%>
DEFAULT rancheros

LABEL rancheros
  KERNEL <%= @kernel -%> rancher.state.dev=LABEL=RANCHER_STATE rancher.state.autoformat=[/dev/xvda] rancher.cloud_init.datasources=['url:<%= foreman_url('provision')-%>']
  APPEND initrd=<%= @initrd -%>```

**Comment by selmison** [comment]
Sample Provision template used in my environment:

```#cloud-config
<%#
kind: provision
name: RancherOS provision
-%>
hostname: <%= @host.name %>
rancher:
  network:
  dns:
    nameservers:
    - <%= @host.subnet.dns_primary %>
    - <%= @host.subnet.dns_secondary %>
    search:
    - <%= @host.domain %>
    interfaces:
      eth*:
        dhcp: false
      eth0:
        address: <%= @host.ip -%>/<%= @host.subnet.cidr %>
        gateway: <%= @host.subnet.gateway %>
  password: 123456```

**Comment by selmison** [comment]
I use [https://github.com/rancher/os/releases/download/v$version](https://github.com/rancher/os/releases/download/v$version) as medium path.

**Comment by lzap** [comment]
Thanks, please add medium path to `db/seeds.d/100-installation_media.rb` and file new PR into https://github.com/theforeman/community-templates so we can test all in once.

Also add icon into `app/assets/images` - you can use original icon in case it is open licensed (e.g. creative commons). If license cannot be secured or unclear, please use dummy icons from https://github.com/lzap/foreman-dummy-icons

Any hints how to test the workflow? This is the first time I will be installing this OS.

**Comment by lzap** [comment]
Also I am missing one important step. At the end of provisioning, we need to call home and tell Foreman to unset "build" flag for a host, otherwise it will end up in endless loop. You can do this with

```yaml
<%# Contact Foreman to confirm instance is built -%>
phone_home:
  url: <%= foreman_url('built') %>
  post: []
  tries: 10
```

File: db/seeds.d/100-installation_media.rb [hunk 1]
```diff
@@ -3,15 +3,15 @@
...
+    { :name => "Debian mirror",        :os_family => "Debian",    :path => "http://ftp.debian.org/debian" },
+    { :name => "Fedora mirror",        :os_family => "Redhat",    :path => "http://dl.fedoraproject.org/pub/fedora/linux/releases/$major/Server/$arch/os/" },
+    { :name => "Fedora Atomic mirror", :os_family => "Redhat",    :path => "http://dl.fedoraproject.org/pub/alt/atomic/stable/Cloud_Atomic/$arch/os/" },
+    { :name => "FreeBSD mirror",       :os_family => "Freebsd",   :path => "http://ftp.freebsd.org/pub/FreeBSD/releases/$arch/$version-RELEASE/" },
+    { :name => "OpenSUSE mirror",      :os_family => "Suse",      :path => "http://download.opensuse.org/distribution/leap/$version/repo/oss", :operatingsystems => os_suse },
+    { :name => "Ubuntu mirror",        :os_family => "Debian",    :path => "http://archive.ubuntu.com/ubuntu" },
+    { :name => "CoreOS mirror",        :os_family => "Coreos",    :path => "http://$release.release.core-os.net" }
+    { :name => "RancherOS mirror",     :os_family => "Rancheros", :path => "https://github.com/rancher/os/releases/download/v$version" }
```
Comment (lzap) [inline db/seeds.d/100-installation_media.rb:14]: Please avoid reformatting dozen of lines just because one line change. Thanks.

**Comment by lzap** [comment]
Hello, the `RancherOS.png` image must be same size as other icons, see the others. Also, we cannot accept any work without license, please provide licensing info and links that will confirm the license.

File: db/seeds.d/100-installation_media.rb [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
+    { :name => "Debian mirror",        :os_family => "Debian",     :path => "http://ftp.debian.org/debian" },
+    { :name => "Fedora mirror",        :os_family => "Redhat",     :path => "http://dl.fedoraproject.org/pub/fedora/linux/releases/$major/Server/$arch/os/" },
+    { :name => "Fedora Atomic mirror", :os_family => "Redhat",     :path => "http://dl.fedoraproject.org/pub/alt/atomic/stable/Cloud_Atomic/$arch/os/" },
+    { :name => "FreeBSD mirror",       :os_family => "Freebsd",    :path => "http://ftp.freebsd.org/pub/FreeBSD/releases/$arch/$version-RELEASE/" },
+    { :name => "OpenSUSE mirror",      :os_family => "Suse",       :path => "http://download.opensuse.org/distribution/leap/$version/repo/oss", :operatingsystems => os_suse },
+    { :name => "Ubuntu mirror",        :os_family => "Debian",     :path => "http://archive.ubuntu.com/ubuntu" },
+    { :name => "CoreOS mirror",        :os_family => "Coreos",     :path => "http://$release.release.core-os.net" }
+    { :name => "RancherOS mirror",     :os_family => "Rancheros",  :path => "https://github.com/rancher/os/releases/download/v$version" }
```
[the remark this answers is not in the document]
Comment (lzap) [reply db/seeds.d/100-installation_media.rb]: Plus parsing error.

**Comment by lzap** [comment]
Thanks again for your contribution, code looks good. Before I test, I need:

* Fix that seed script (only change one line)
* Create all necessary templates in community-templates and link the PR from here
* Fix the icon size and clear the license, or use stub icon

Then I will start testing

**Review by zozoh94 — APPROVED** [review]

**Comment by mmoll** [comment]
@zozoh94 did you test this?

**Comment by zozoh94** [comment]
Not enough for the moment. I will need this functionnality soon so I forked the repo. I will purpose a pull request with fixes mentionned by Izap above in the coming weeks.

**Comment by zozoh94** [comment]
@lzap  Sorry to ask that but what do you mean by :  clear the license, or use stub icon ?

**Comment by selmison** [comment]
@zozoh94 

Maybe he just wants a link with information about the license of the rancheros:
 [https://github.com/rancher/os/blob/master/LICENSE](https://github.com/rancher/os/blob/master/LICENSE)

About stub icon:
I believe the size of the image I comitted is larger than the adopted standard

**Comment by zozoh94** [comment]
@selmison I did it here [](https://github.com/zozoh94/foreman/tree/rancher_support) tomorrow I will test the code, correct it and do a pull request. Thank you for your job :+1:

**Comment by lzap** [comment]
Yeah I want to clear icon license, the problem is that companies do not like their brands to be floating around, that's why Fedora or CentOS is not named RHEL and have different logo. There are usually some policies about "fair use" etc, IANAL. Created a thread on their forums:

https://forums.rancher.com/t/rancher-os-logo-license/9879

Let's see. In the meantime, can you resample it to the proper size? Thanks.

**Comment by vincent99** [comment]
We don't have a problem with using our logo, but the one in this PR is outdated (no `®`) and looks to have been maybe scaled from another bitmap.  

There are vector versions on the page below, if you can use those directly. Otherwise you can rasterize into the desired size.

There is not a stacked version exactly like one you have here; if the one line version is no good, let me know and I'll have our designer make it.

http://rancher.com/brand-guidelines/

**Comment by zozoh94** [comment]
@vincent99 I used in my pull request the SVG which is on the page you linked. I rasterized it to 21x21px (of course at this size the ® is invisible). I don't know if I shoud and how to use the SVG directly...

**Comment by lzap** [comment]
@vincent99 thanks for confirmation. We can go ahead and use the official logo then, hopefully it looks good enough in this size.

@zozoh94 please rebase the PR, I still see the old logo which is 400x377 in size. Since Foreman is used in enterprise world quite widely (really old browsers), I'd vote for (transparent) PNG version.

Thanks all!

**Comment by lzap** [comment]
At this scale I'd recommend to use just the icon with no text, so it's recognizable. Pick either blue or black variant. Here is the scaled version to 21x21. Does not look super good, maybe you want to get your designers to be involved @zozoh94 ?

rancheros

**Comment by zozoh94** [comment]
@lzap I did it in my pull request [](https://github.com/theforeman/foreman/pull/5361)

**Comment by zozoh94** [comment]
@lzap that's what I did in my pull request with the blue one. I think it's enough, the logo is recognizable. Anyway it's very tiny in the OS list ! 
I will rebase develop in my PR(https://github.com/theforeman/foreman/pull/5361)

**Comment by mmoll** [comment]
closing in favour of GH-5361

**Pull request closed without merging**